### PR TITLE
<DRAFT>Create warehouse pool

### PR DESCRIPTION
**Pull request by anshuldata**
# Pull Request Template

## Description

Please include a summary of the change including relevant motivation and context.

## Checklist:

- [ ] I have verified that my change is functionally correct (unit tests are great).
- [ ] Queries against newly-added, user-facing objects can be expected to complete in 1's of seconds.
- [ ] My change will not have a substantial increase the user's cost to have the app installed (did you add any new tasks?).

### CODE REVIEW

File: app/crud/warehouse_pools.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                params=(self.name,),
+            ).collect()[0][0]
+            assert count == 0, "A warehouse pool with this name already exists."
+            super().write(txn)
+
+    @validator("name", allow_reuse=True)
+    def verify_name(cls, v):
+        if not v:
+            raise ValueError("Name is required")
+        assert isinstance(v, str)
```
Comment (pravindra) [inline app/crud/warehouse_pools.py:78]: can we also check that it is a valid unquoted identifier ? We are also using this as a prefix/suffix in generating warehouse names - so, it's less confusing to have simpler names.

File: bootstrap/013_warehouse_pools.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+AS
+$$
+from ml.warehouse_pool import compute_set_of_warehouses
+from crud.base import transaction
+from crud.warehouse_pools import Warehouse, WarehousePools
+def create_warehouse_pool(bare_session, pool_name: str, max_concurrent_credits: int, allowed_roles: list[str], snowflake_tag: dict, target_label: str):
+    warehouses = []
+    dataframe = compute_set_of_warehouses('Medium', max_concurrent_credits)
+    # TODO: get random number, secrets/random package wasn't working
+    rand_prefix = 123458
```
Comment (pravindra) [inline bootstrap/013_warehouse_pools.sql:19]: if this is not working, should we move the logic to a python fn ? or, compute_set_of_warehouses() ?

File: bootstrap/013_warehouse_pools.sql [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    imports = ('{{stage}}/python/crud.zip', '{{stage}}/python/ml.zip')
+    EXECUTE AS OWNER
+AS
+$$
+from ml.warehouse_pool import compute_set_of_warehouses
+from crud.base import transaction
+from crud.warehouse_pools import Warehouse, WarehousePools
+def create_warehouse_pool(bare_session, pool_name: str, max_concurrent_credits: int, allowed_roles: list[str], snowflake_tag: dict, target_label: str):
+    warehouses = []
+    dataframe = compute_set_of_warehouses('Medium', max_concurrent_credits)
```
Comment (pravindra) [inline bootstrap/013_warehouse_pools.sql:17]: default wh size should be a param too

File: bootstrap/013_warehouse_pools.sql [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    # TODO: get random number, secrets/random package wasn't working
+    rand_prefix = 123458
+    for row in dataframe.itertuples():
+        # TODO: this can be parallelized
+        rand_prefix += 1
+        # TODO: get random number, secrets wasn't working'
+        wh_name = f'SD_{pool_name}_{rand_prefix}'
+        wh_size = row.warehouse_size
+        autoscale_min = 1
+        autoscale_max = row.num_clusters
```
Comment (pravindra) [inline bootstrap/013_warehouse_pools.sql:27]: can we do a hack to set this to 1 if autoscaling is disabled in the account ?

File: bootstrap/013_warehouse_pools.sql [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                     snowflake_tag)
+        for role in allowed_roles:
+            bare_session.sql("GRANT USAGE ON WAREHOUSE ? TO ROLE ?", wh_name, role)
+
+        warehouses.append(Warehouse(name=wh_name, size=wh_size, autoscale_min=autoscale_min,
+                                    autoscale_max=autoscale_max))
+
+    new_pool = WarehousePools.parse_obj(dict(
+            name=pool_name,
+            warehouses=warehouses,
```
Comment (pravindra) [inline bootstrap/013_warehouse_pools.sql:38]: save the other params too : credits, roles, defaultWarehouseSize, .. for showing in the UI

File: app/crud/warehouse_pools.py (hunk 1 above)
Comment (anshuldata) [reply app/crud/warehouse_pools.py:92]: fixed it

File: bootstrap/013_warehouse_pools.sql (hunk 2 above)
Comment (anshuldata) [reply bootstrap/013_warehouse_pools.sql:44]: fixed it. defined a function and called it

File: bootstrap/013_warehouse_pools.sql (hunk 3 above)
Comment (anshuldata) [reply bootstrap/013_warehouse_pools.sql]: added

File: bootstrap/013_warehouse_pools.sql (hunk 4 above)
Comment (anshuldata) [reply bootstrap/013_warehouse_pools.sql:51]: I need to figure out how to do this. I will do it as a separate PR

File: bootstrap/013_warehouse_pools.sql (hunk 5 above)
Comment (anshuldata) [reply bootstrap/013_warehouse_pools.sql:63]: added

**Comment by anshuldata** [comment]
Closing this, will raise it against new repo

**Pull request closed without merging**